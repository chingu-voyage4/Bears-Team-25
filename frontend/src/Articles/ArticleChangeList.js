import React from 'react';
import PropTypes from 'prop-types';

import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableSortLabel from '@material-ui/core/TableSortLabel';

import actions from './actions';

import { LinkCell, TableRow, TableWrapper, Wrapper } from './styled';

const REQUESTER = 'requester';

const columnData = [
  { id: 'topic', numeric: false, label: 'Topic' },
  { id: 'sub_topic', numeric: false, label: 'Sub Topic' },
  { id: 'title', numeric: false, label: 'Title' },
  { id: REQUESTER, numeric: false, label: 'Requester' },
];

export default class ArticleChangeList extends React.Component {
  static propTypes = {
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }).isRequired,
    location: PropTypes.shape({
      pathname: PropTypes.string.isRequired,
    }).isRequired,
  };

  state = {
    data: [],
    order: 'asc',
  };

  componentDidMount = () => {
    this.fetchData()
      .then(res => {
        console.log('article change list response:', res);
        const data = res.map(request => ({
          ...request,
          title: request.article.title,
          topic: request.article.topic.name,
          sub_topic: request.article.sub_topic
            ? request.article.sub_topic.name
            : '',
        }));
        this.setState({ data });
      })
      // eslint-disable-next-line no-console
      .catch(e => console.error('mounted error:', e));
  };

  // eslint-disable-next-line class-methods-use-this
  fetchData() {
    return actions.getArticleChangeRequestList();
  }

  handleSort = id => {
    const { data, sortedBy } = this.state;
    let { order } = this.state;

    if (id !== sortedBy) {
      order = 'desc';
    }

    if (id === REQUESTER) {
      data.sort((a, b) => {
        if (a[id].username.toLowerCase() < b[id].username.toLowerCase())
          return order === 'asc' ? 1 : -1;
        if (a[id].username.toLowerCase() > b[id].username.toLowerCase())
          return order === 'asc' ? -1 : 1;
        return 0;
      });
    } else {
      data.sort((a, b) => {
        if (a[id].toLowerCase() < b[id].toLowerCase())
          return order === 'asc' ? 1 : -1;
        if (a[id].toLowerCase() > b[id].toLowerCase())
          return order === 'asc' ? -1 : 1;
        return 0;
      });
    }

    const newOrder = order === 'asc' && sortedBy === id ? 'desc' : 'asc';

    this.setState(() => ({
      data,
      order: newOrder,
      sortedBy: id,
    }));
  };

  handleClick = (e, articleId, requesterId) => {
    e.preventDefault();

    const { name } = e.target.dataset;

    if (name === REQUESTER) {
      this.props.history.push(`/users/${requesterId}`);
    } else {
      this.props.history.push(`${this.props.location.pathname}/${articleId}`);
    }
  };

  render() {
    const { data, order, sortedBy } = this.state;
    return (
      <Wrapper>
        <TableWrapper>
          <Table>
            <TableHead>
              <TableRow>
                {columnData.map(column => (
                  <TableCell key={column.id} numeric={column.numeric}>
                    <TableSortLabel
                      active={column.id === sortedBy}
                      direction={order}
                      onClick={() => this.handleSort(column.id)}
                    >
                      {column.label}
                    </TableSortLabel>
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {data.map(d => (
                <TableRow
                  key={d._id}
                  onClick={e =>
                    this.handleClick(e, d.article._id, d.requester._id)
                  }
                >
                  <TableCell>{d.topic}</TableCell>
                  <TableCell>{d.sub_topic}</TableCell>
                  <TableCell>{d.title}</TableCell>
                  <LinkCell data-name={REQUESTER}>
                    {d.requester.username}
                  </LinkCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableWrapper>
      </Wrapper>
    );
  }
}
