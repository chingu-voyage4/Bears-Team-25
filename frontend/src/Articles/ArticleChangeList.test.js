import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import ArticleChangeList from './ArticleChangeList';

const requests = [
  {
    _id: '5b133cf0ebd521b1d4a5144a',
    article: {
      id: '5ae27568463f344afc3ac08e',
      title: 'test article',
      topic: { id: 'topic_id', name: 'test topic' },
      sub_topic: { id: 'sub_topic_id', name: 'test sub topic' },
      content: 'TBD',
    },
    diff: [
      {
        count: 3,
        value: 'TBD',
      },
      {
        count: 5,
        added: true,
        value: ' test',
      },
    ],
    status: 'pending',
    requester: { id: '5ae34a775cfd8b560525c60c', username: 'test user' },
  },
];

const props = {
  history: {
    push: jest.fn(),
  },
  location: {
    pathname: '/requests',
  },
};

describe('Article change request list', () => {
  it('should match snapshot', () => {
    jest
      .spyOn(ArticleChangeList.prototype, 'fetchData')
      .mockImplementation(() => new Promise(resolve => resolve(requests)));

    const comp = shallow(<ArticleChangeList {...props} />);
    return Promise.resolve().then(() => {
      comp.update();
      expect(toJSON(comp)).toMatchSnapshot();
    });
  });
});
