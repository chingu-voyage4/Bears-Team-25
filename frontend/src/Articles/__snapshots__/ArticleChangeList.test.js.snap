// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article change request list should match snapshot 1`] = `
<Wrapper>
  <Styled(WithStyles(Paper))>
    <WithStyles(Table)>
      <WithStyles(TableHead)>
        <Styled(WithStyles(TableRow))>
          <WithStyles(TableCell)
            key="topic"
            numeric={false}
          >
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
              onClick={[Function]}
            >
              Topic
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="sub_topic"
            numeric={false}
          >
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
              onClick={[Function]}
            >
              Sub Topic
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="title"
            numeric={false}
          >
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
              onClick={[Function]}
            >
              Title
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="requester"
            numeric={false}
          >
            <WithStyles(TableSortLabel)
              active={false}
              direction="asc"
              onClick={[Function]}
            >
              Requester
            </WithStyles(TableSortLabel)>
          </WithStyles(TableCell)>
        </Styled(WithStyles(TableRow))>
      </WithStyles(TableHead)>
      <WithStyles(TableBody)>
        <Styled(WithStyles(TableRow))
          key="5b133cf0ebd521b1d4a5144a"
          onClick={[Function]}
        >
          <WithStyles(TableCell)>
            test topic
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            test sub topic
          </WithStyles(TableCell)>
          <WithStyles(TableCell)>
            test article
          </WithStyles(TableCell)>
          <Styled(WithStyles(TableCell))
            data-name="requester"
          >
            test user
          </Styled(WithStyles(TableCell))>
        </Styled(WithStyles(TableRow))>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </Styled(WithStyles(Paper))>
</Wrapper>
`;
